corgi/core/models.py

```diff
@@ -930,12 +932,27 @@ def latest_components(
         include: bool = True,
     ) -> "ComponentQuerySet":
         """Return components from latest builds across all product streams."""
-        distinct_components = (
-            self.values_list("namespace", "name", "arch").order_by().distinct().iterator()
+        nevras_by_components = self.values_list("namespace", "name", "arch").annotate(
+            nevras=JSONBAgg(
+                RawSQL(
+                    "json_build_object('nevra', nevra, 'epoch', epoch, 'version', core_component.version, 'release', release)",
+                    (),
+                )
+            )
         )
         query = Q()
-        for namespace, name, arch in distinct_components:
-            latest_nevra = self.filter_latest_nevra_by_distinct_component(namespace, name, arch)
+        for _, _, _, nevra_data in nevras_by_components:
+            nevras = []
+            for nevra in nevra_data:
+                nevras.append(
+                    (
+                        nevra["nevra"],
+                        nevra["epoch"],
+                        nevra["version"],
+                        nevra["release"],
+                    )
+                )
+            latest_nevra = sorted(nevras, key=functools.cmp_to_key(self._compare_packages))[-1]
```

Comment:
Let's just make the list we want directly, instead of appending to it:
```suggestion
            nevras = tuple(
                (nevra["nevra"], nevra["epoch"], nevra["version"], nevra["release"])
                for nevra in nevra_data
            )
            latest_nevra = sorted(nevras, key=functools.cmp_to_key(self._compare_packages))[-1]
```

We can also avoid building the whole list in memory and calling sorted(), which will cause OoM if the number of items to sort is very large:
```suggestion
            latest_nevra = nevra_data[0]
            for current_nevra in nevra_data:
                # if the current nevra is newer than the "latest" nevra we already have
                if self._compare_packages(current_nevra, latest_nevra):
                    latest_nevra = current_nevra
```

Reply:
I don't think the first suggestion here will make any difference to the memory consumption, and you seem to be missing the type cast of the epoch value to string. I thought this loop was complicated enough to warrant using a traditional for loop.

I like the 2nd option, I'll update the loop to compare during the iterator to avoid building the list.

Reply:
Yes, I was commenting on an older commit (looking at each commit individually to understand better) so the diff here doesn't have the epoch-to-str cast. The first suggestion is just a strong style preference. The second one will avoid out of memory issues. Thanks for fixing!